### ISSUE TITLE
Segmentation fault deserializing a network message - overflow in uniform_typ_info_map

### ISSUE THREAD

**Issue by nq-ebaratte** [actor-framework/actor-framework#388]
TL;DR: `uti_impl::by_type_nr()` does not perform upper bounds checking; it probably should, since the values may come from the network, ie. any (garbage) value could come in.

-- 

I randomly get a segfault when deserializing a message coming from the network (message to a remote actor).
(Note: the incoming data seems to be garbage, the reason of which I am still investigating)

The stacktrace I get is:

    (gdb) where
    #0  0x0000000000000050 in ?? ()
    #1  0x0000000000ded757 in caf::message::deserialize (this=0x7fffd622c860, source=...)
        at /home/etienne/actor-framework/libcaf_core/src/message.cpp:70
    #2  0x0000000000ec3bd0 in caf::io::basp::instance::handle (this=0x7fffe0033af8, dm=..., hdr=..., is_payload=true)
        at /home/etienne/actor-framework/libcaf_io/src/basp.cpp:467
    #3  0x0000000000e8f17d in caf::io::basp_broker::<lambda(caf::io::new_data_msg&)>::operator()(caf::io::new_data_msg &) const (__closure=0x7fffb8001a30, msg=...)
        at /home/etienne/actor-framework/libcaf_io/src/basp_broker.cpp:448
    #4  0x0000000000ea78a5 in caf::lfinvoker<true, caf::io::basp_broker::make_behavior()::<lambda(caf::io::new_data_msg&)> >::operator()<caf::io::new_data_msg&>(void) (this=0x7fffd622ca20)
        at /home/etienne/actor-framework/libcaf_core/caf/match_case.hpp:134

    [snip]
    
    #14 0x0000000000e630b0 in caf::io::network::stream<caf::io::network::default_socket>::handle_event (this=0x7fffb800b930, op=caf::io::network::read)
        at /home/etienne/actor-framework/libcaf_io/caf/io/network/default_multiplexer.hpp:510
    #15 0x0000000000e5834f in caf::io::network::default_multiplexer::handle_socket_event (this=0x7fffe0032cd0, fd=142, mask=1, ptr=0x7fffb800b930) 
        at /home/etienne/actor-framework/libcaf_io/src/default_multiplexer.cpp:686
    #16 0x0000000000e578b9 in caf::io::network::default_multiplexer::run (this=0x7fffe0032cd0)
        at /home/etienne/actor-framework/libcaf_io/src/default_multiplexer.cpp:326
    #17 0x0000000000e66994 in caf::io::middleman::<lambda()>::operator()(void) const (__closure=0x7fffe0033088)
        at /home/etienne/actor-framework/libcaf_io/src/middleman.cpp:262

Relevant source (message.cpp:68)

    68     void message::deserialize(deserializer& source) {
    69       auto uti = source.begin_object();
    70       auto uval = uti->create();

According to gdb, the uti variable is garbage; it comes from the following method (binary_deserializer.cpp:201)

    const uniform_type_info* binary_deserializer::begin_object() {
      auto uti_map = detail::singletons::get_uniform_type_info_map();
      detail::uniform_type_info_map::pointer uti;
      uint16_t nr;
      pos_ = read_range(pos_, end_, nr);
      if (nr) {
        uti = uti_map->by_type_nr(nr);
      } else {
      // [...]
      }
      return uti;

The value of `nr`read from the buffer is `10792` which I believe is not an acceptable type index: 

      virtual pointer by_type_nr(uint16_t nr) const {
        CAF_ASSERT(nr > 0);
        return builtin_types_[nr - 1];
      }

No upper bounds checking is done on the value - at least `builtin_types_.at(nr - 1)` would throw an exception instead of overflowing, that could be caught above.

**Comment by Neverlord** [comment]
> (Note: the incoming data seems to be garbage, the reason of which I am still investigating)

Could you find the steps to reproduce this?

**Comment by nq-ebaratte** [comment]
Not really: it occurred while running heavy load tests, with thousands of connections between different nodes, and a few thousands messages per seconds. Among these, one gets eventually garbled. 

A few thoughts I had:
 * I tried to enable logging, but I filled the filesystem before reproducing the issue :(
 * I carefully checked that every message type is announced (btw, is there a way to guarantee it, some static_assert or even assert ?)
 * it may be related to the amount of opened connections (I had to increase ulimit -n)

I am still investigating on my side.

**Comment by Neverlord** [comment]
> I tried to enable logging, but I filled the filesystem before reproducing the issue :(

Too bad.

> I carefully checked that every message type is announced (btw, is there a way to guarantee it, some static_assert or even assert ?)

That's something we could introduce with 0.15. CAF could reject any non-serializable type per default, allowing users to explicitly whitelisting "unsafe" types (for messages that are guaranteed to never leave an actor system, i.e., when using CAF for concurrency only).

> I am still investigating on my side.

Good luck! Are you reproducing this issue with a test case you can share?

**Comment by nq-ebaratte** [comment]
Still not able to isolate a test case exhibiting the issue.

Some additional info, that may or may not be related:
 * the largest number of remote connections between actors on my system comes from remote group subscriptions 
 * it occurs once the server is heavily loaded, specially in terms of open connections (I have litteraly thousands of opened sockets; hence #390); sharing the remote_actor used to connect different remote groups hosted on the same server drastically reduces this number and fixes (or hides ?) the issue. 
 * one of the clients returns a "disconnect during handshake" error at the time of the crash

**Comment by ebaratte** [comment]
We haven't seen this behavior for several months. Maybe it was some memory mismanagement on our side. Shall we close it ?

**Comment by Neverlord** [comment]
> Shall we close it ?

I'd say yes. Feel free to re-open if something similar pops up again.
